### PR TITLE
Experimental Istio gateway functionality

### PR DESCRIPTION
**Pull request by liamawhite**
Signed-off-by: Liam White <liam@tetrate.io>

### CODE REVIEW

**Comment by liamawhite** [comment]
Ready for review. Tests to follow!

File: pkg/binary/envoy/controlplane/istio.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	"istio.io/istio/pkg/config/mesh"
+)
+
+const (
+	defaultControlplane   = "istio-pilot:15010"
+	initialEpochBootstrap = "envoy-rev1.json"
+)
+
+// Istio tells GetEnvoy that it's using Istio for xDS and should bootstrap accordingly
+var Istio = func(r *envoy.Runtime) {
```
Comment (ZackButcher) [inline pkg/binary/envoy/controlplane/istio.go]: I'd rather see this be `func Istio(r *envoy.Runtime)` - at your usage site it's identical, but that way you can't re-assign this function.

File: pkg/binary/envoy/controlplane/istio.tmpl.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+
+// TODO: (maybe?) Refactor the Istio write.Bootstrap upstream so we can pass it a byte slice instead
+func writeIstioTemplate(path string) error {
+	if err := os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(path, []byte(istioBootStrapTemplate), 0644)
+}
+
+var istioBootStrapTemplate = `{
```
[the remark this answers is not in the document]
Comment (lizan) [reply pkg/binary/envoy/controlplane/istio.tmpl.go]: The change to stats_config is relatively frequent (to adjust for new stats added / remove useless ones).

File: ci/install-lint [hunk 3]
```diff
@@ -4,4 +4,4 @@
 wget -c https://github.com/liamawhite/licenser/releases/download/v0.5.1/licenser_0.5.1_Linux_x86_64.tar.gz -O - | sudo tar -xz -C /usr/bin
 
 # Install golangci-lint
-curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s -- -b $(go env GOPATH)/bin v1.17.1
+curl -sfL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sudo sh -s -- -b /usr/bin v1.17.1
```
[the remark this answers is not in the document]
Comment (lizan) [reply ci/install-lint]: This script is specific to CI so I guess it is fine

File: pkg/binary/envoy/controlplane/istio.tmpl.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liamawhite) [reply pkg/binary/envoy/controlplane/istio.tmpl.go]: My thinking was that I dont want users to have to ship files around with the CLI, I want it to "just work".

We could pull the bootstrap in at runtime but I'm not sure that's preferable.

I could add a test that ensures it matches the latest Istio version?

File: ci/install-lint (hunk 3 above)
[the remark this answers is not in the document]
Comment (liamawhite) [reply ci/install-lint]: Yeah, the script is not meant for local consumption.

**Pull request merged**